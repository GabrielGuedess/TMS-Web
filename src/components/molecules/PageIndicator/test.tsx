import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

import { PageIndicator } from '.';

const onChangeIndex = jest.fn();
const onChangeWithEdited = jest.fn();

describe('<PageIndicator />', () => {
  beforeEach(() => {
    onChangeIndex.mockClear();
    onChangeWithEdited.mockClear();
  });

  it('should render the page indicator correctly', () => {
    const { container } = render(
      <PageIndicator
        length={5}
        activeIndex={0}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be able change page with page indicator indicator button previous when is not index 0', async () => {
    render(
      <PageIndicator
        length={5}
        activeIndex={3}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
      />,
    );

    const previousButton = screen.getByRole('button', { name: 'Previous' });

    await userEvent.click(previousButton);

    expect(onChangeIndex).toHaveBeenCalled();
  });

  it('should not be able change page with page indicator button previous when is not index 0 and has edited', async () => {
    render(
      <PageIndicator
        length={5}
        activeIndex={3}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
        hasEdited
      />,
    );

    const previousButton = screen.getByRole('button', { name: 'Previous' });

    await userEvent.click(previousButton);

    expect(onChangeIndex).not.toHaveBeenCalled();
    expect(onChangeWithEdited).toHaveBeenCalled();
  });

  it('should be able change page with page indicator button next when is not loading', async () => {
    render(
      <PageIndicator
        length={5}
        activeIndex={0}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
      />,
    );

    const nextButton = screen.getByRole('button', { name: 'Next' });

    await userEvent.click(nextButton);

    expect(onChangeIndex).toHaveBeenCalled();
  });

  it('should be not able change page with page indicator button next when is loading', async () => {
    render(
      <PageIndicator
        length={5}
        activeIndex={0}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
        isLoading
      />,
    );

    const nextButton = screen.getByRole('button', { name: 'Next' });

    await userEvent.click(nextButton);

    expect(onChangeIndex).not.toHaveBeenCalled();
  });

  it('should not be able change page with page indicator button next when is not loading and has edited', async () => {
    render(
      <PageIndicator
        length={5}
        activeIndex={0}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
        hasEdited
      />,
    );

    const nextButton = screen.getByRole('button', { name: 'Next' });

    await userEvent.click(nextButton);

    expect(onChangeIndex).not.toHaveBeenCalled();
    expect(onChangeWithEdited).toHaveBeenCalled();
  });

  it('should be able change page one with page indicator when is not loading', async () => {
    render(
      <PageIndicator
        length={5}
        activeIndex={3}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
      />,
    );

    const nextButton = screen.getByRole('button', { name: 'First Page' });

    await userEvent.click(nextButton);

    expect(onChangeIndex).toHaveBeenCalledWith(0);
  });

  it('should not be able change page one with page indicator when is not loading and has edited', async () => {
    render(
      <PageIndicator
        length={5}
        activeIndex={3}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
        hasEdited
      />,
    );

    const nextButton = screen.getByRole('button', { name: 'First Page' });

    await userEvent.click(nextButton);

    expect(onChangeIndex).not.toHaveBeenCalled();
    expect(onChangeWithEdited).toHaveBeenCalled();
  });

  it('should be able change page one with page indicator button number when is not loading', async () => {
    render(
      <PageIndicator
        length={5}
        activeIndex={3}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
      />,
    );

    const numberButton = screen.getByRole('button', { name: 'Button 2' });

    await userEvent.click(numberButton);

    expect(onChangeIndex).toHaveBeenCalledWith(1);
  });

  it('should not be able change page one with page indicator button number when is not loading and has edited', async () => {
    render(
      <PageIndicator
        length={5}
        activeIndex={3}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
        hasEdited
      />,
    );

    const numberButton = screen.getByRole('button', { name: 'Button 2' });

    await userEvent.click(numberButton);

    expect(onChangeIndex).not.toHaveBeenCalled();
    expect(onChangeWithEdited).toHaveBeenCalled();
  });

  it('should be able change page one with page indicator button number last item when is not loading', async () => {
    render(
      <PageIndicator
        length={20}
        activeIndex={3}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
      />,
    );

    const numberButton = screen.getByRole('button', { name: 'Last Page' });

    await userEvent.click(numberButton);

    expect(onChangeIndex).toHaveBeenCalledWith(19);
  });

  it('should not be able change page one with page indicator button number last item when is not loading and has edited', async () => {
    render(
      <PageIndicator
        length={20}
        activeIndex={3}
        onChangeIndex={onChangeIndex}
        onChangeWithEdited={onChangeWithEdited}
        hasEdited
      />,
    );

    const numberButton = screen.getByRole('button', { name: 'Last Page' });

    await userEvent.click(numberButton);

    expect(onChangeIndex).not.toHaveBeenCalled();
    expect(onChangeWithEdited).toHaveBeenCalled();
  });
});
