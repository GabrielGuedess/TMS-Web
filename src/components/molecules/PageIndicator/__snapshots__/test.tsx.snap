// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageIndicator /> should render the page indicator correctly 1`] = `
<div
  class="flex gap-2"
>
  <button
    class="flex h-9 items-center justify-center rounded-md bg-dark-500/[.08] px-[0.745rem] text-[0.938rem] leading-none text-gray-900 outline-primary-400 transition-all dark:bg-shark-950 dark:text-dark-300 opacity-50 cursor-not-allowed"
    disabled=""
    type="button"
  >
    Previous
  </button>
  <button
    aria-label="Button 1"
    class="flex size-9 items-center justify-center rounded-md text-[0.938rem] leading-none outline-primary-400 transition-all bg-primary-400 text-white"
    type="button"
  >
    <span
      aria-label="Current Page Active"
      class="absolute"
    >
      1
    </span>
  </button>
  <button
    aria-label="Button 2"
    class="flex size-9 items-center justify-center rounded-md text-[0.938rem] leading-none outline-primary-400 transition-all text-shark-950 dark:text-dark-300"
    type="button"
  >
    <span
      aria-label="Current Page"
      class="absolute"
    >
      2
    </span>
  </button>
  <button
    aria-label="Button 3"
    class="flex size-9 items-center justify-center rounded-md text-[0.938rem] leading-none outline-primary-400 transition-all text-shark-950 dark:text-dark-300"
    type="button"
  >
    <span
      aria-label="Current Page"
      class="absolute"
    >
      3
    </span>
  </button>
  <div
    class="hidden md:flex"
  >
    <span
      class="flex items-end text-gray-900 transition-all dark:text-dark-300"
    >
      ...
    </span>
    <button
      aria-label="Last Page"
      class="flex h-9 items-center justify-center rounded-md px-[1.1rem] text-[0.938rem] leading-none outline-primary-400 transition-all text-gray-900 dark:text-dark-300"
      type="button"
    >
      5
    </button>
  </div>
  <button
    class="flex h-9 items-center justify-center rounded-md bg-dark-500/[.08] px-[0.745rem] text-[0.938rem] leading-none text-gray-900 outline-primary-400 transition-all dark:bg-shark-950 dark:text-dark-300 cursor-pointer hover:bg-darkGray/[0.16]"
    type="button"
  >
    Next
  </button>
</div>
`;
